### PR TITLE
【Prim】Reshape, transpose, cast vjp

### PR DESCRIPTION
**Pull request by JiabinYang**
### PR types

Others
### PR changes

Others
### Describe

This PR support Reshape, Transpose, Cast ops' vjp...

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/prim/prim/vjp/static/test_comp_reshape_grad.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        (
+            np.random.rand(10, 1, 10),
+            [10, 10],
+            np.random.rand(10, 10),
+            np.float32,
+        ),
+        (np.random.rand(2, 60), [12, 10], np.random.rand(12, 10), np.float32),
+    ],
+)
+class TestSqrtGradComp(unittest.TestCase):
```
Comment (xiaoguoguo626807) [inline python/paddle/fluid/tests/unittests/prim/prim/vjp/static/test_comp_reshape_grad.py]: change name to reshape

File: python/paddle/fluid/tests/unittests/prim/prim/vjp/eager/test_comp_eager_transpose_grad.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            np.random.rand(2, 3, 4, 5, 6, 1),
+            [4, 2, 3, 1, 0, 5],
+            np.random.rand(6, 4, 5, 3, 2, 1),
+            np.float32,
+        ),
+        # (np.random.rand(),
+        #  [],
+        # np.random.rand(),
+        # np.float32),
+    ],
```
Comment (xiaoguoguo626807) [inline python/paddle/fluid/tests/unittests/prim/prim/vjp/eager/test_comp_eager_transpose_grad.py:65]: why delete this case

File: python/paddle/fluid/tests/unittests/prim/prim/vjp/eager/test_comp_eager_reshape_grad.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        (
+            np.random.rand(10, 1, 10),
+            [10, 10],
+            np.random.rand(10, 10),
+            np.float32,
+        ),
+        (np.random.rand(2, 60), [12, 10], np.random.rand(12, 10), np.float32),
+    ],
+)
+class TestSqrtGradComp(unittest.TestCase):
```
Comment (xiaoguoguo626807) [inline python/paddle/fluid/tests/unittests/prim/prim/vjp/eager/test_comp_eager_reshape_grad.py]: change name

File: python/paddle/fluid/tests/unittests/prim/prim/vjp/eager/test_comp_eager_cast_grad.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+            np.random.rand(10, 10),
+            np.float64,
+            np.float32,
+        ),
+        (
+            np.random.rand(10, 10),
+            np.random.rand(10, 10),
+            np.float32,
+            np.float32,
+        ),
```
Comment (xiaoguoguo626807) [inline python/paddle/fluid/tests/unittests/prim/prim/vjp/eager/test_comp_eager_cast_grad.py:45]: add test fp16 <-> fp32

File: paddle/fluid/prim/api/composite_backward/composite_backward_api.h [hunk 5]
```diff
@@ -25,6 +25,13 @@ using IntArray =
 //  This function should have as same signature as phi, which defined in
 //  paddle/phi/api/backward/backward_api.h
 template <typename T>
+void cast_grad(const Tensor& out_grad, DataType dtype, Tensor* x_grad) {
+  if (x_grad) {
+    auto res = cast<T>(out_grad, dtype);
+    set_output<T>(res, x_grad);
+  }
+}
```
Comment (xiaoguoguo626807) [inline paddle/fluid/prim/api/composite_backward/composite_backward_api.h:33]: add line

File: paddle/fluid/prim/api/manual_prim/eager_prim_api.cc [hunk 6]
```diff
@@ -33,5 +33,9 @@ Tensor full<Tensor>(const IntArray& shape,
   VLOG(4) << "Eager Prim API full_ad_func call";
   return ::full_ad_func(shape, value, dtype, place);
 }
+template <>
+Tensor cast<Tensor>(const Tensor& x, DataType dtype) {
+  return ::cast_ad_func(x, dtype);
+}
```
Comment (xiaoguoguo626807) [inline paddle/fluid/prim/api/manual_prim/eager_prim_api.cc:39]: why cast can't be generated

File: python/paddle/fluid/tests/unittests/prim/prim/vjp/static/test_comp_cast_grad.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+        return res
+
+    def test_cinn(self):
+        paddle.disable_static()
+        use_cinn = True
+        if isinstance(
+            framework._current_expected_place(), framework.core.CPUPlace
+        ):
+            # TODO(jiabin): CINN will crashed in this case open it when fixed
```
Comment (xiaoguoguo626807) [inline python/paddle/fluid/tests/unittests/prim/prim/vjp/static/test_comp_cast_grad.py:95]: why pass cpu test

**Review by xiaoguoguo626807 — DISMISSED** [review]

File: python/paddle/fluid/tests/unittests/prim/prim/vjp/eager/test_comp_eager_transpose_grad.py (hunk 2 above)
Comment (JiabinYang) [reply python/paddle/fluid/tests/unittests/prim/prim/vjp/eager/test_comp_eager_transpose_grad.py:65]: This is a 0-D tensor case, which is not supported for now.

File: python/paddle/fluid/tests/unittests/prim/prim/vjp/static/test_comp_reshape_grad.py (hunk 1 above)
Comment (JiabinYang) [reply python/paddle/fluid/tests/unittests/prim/prim/vjp/static/test_comp_reshape_grad.py]: done

File: python/paddle/fluid/tests/unittests/prim/prim/vjp/eager/test_comp_eager_reshape_grad.py (hunk 3 above)
Comment (JiabinYang) [reply python/paddle/fluid/tests/unittests/prim/prim/vjp/eager/test_comp_eager_reshape_grad.py]: done

File: paddle/fluid/prim/api/manual_prim/eager_prim_api.cc (hunk 6 above)
Comment (JiabinYang) [reply paddle/fluid/prim/api/manual_prim/eager_prim_api.cc:39]: cast has some special attr which is not supported for now

File: python/paddle/fluid/tests/unittests/prim/prim/vjp/static/test_comp_cast_grad.py (hunk 7 above)
Comment (JiabinYang) [reply python/paddle/fluid/tests/unittests/prim/prim/vjp/static/test_comp_cast_grad.py:95]: cinn will crash on cpu

File: python/paddle/fluid/tests/unittests/prim/prim/vjp/eager/test_comp_eager_cast_grad.py (hunk 4 above)
Comment (JiabinYang) [reply python/paddle/fluid/tests/unittests/prim/prim/vjp/eager/test_comp_eager_cast_grad.py:45]: done

**Review by xiaoguoguo626807 — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**